### PR TITLE
Modify a NSMutableAttributedString directly

### PR DESCRIPTION
**Pull request by colasbd**

### CODE REVIEW

**Comment by alehed** [comment]
Do you have a specific use case in mind?
I'm not quite sure if always adding attributes without without ever removing them is doing what you intend to do.

File: SyntaxKit/AttributedParser.swift [hunk 1]
```diff
@@ -43,6 +43,16 @@ open class AttributedParser: Parser {
         }
     }
 
+    open func addAttributes(to mutableString: NSMutableAttributedString) -> Void {
```
Comment (alehed) [inline SyntaxKit/AttributedParser.swift]: We don't need the void there

Comment (colasbd) [reply SyntaxKit/AttributedParser.swift]: Ok (I'm learning Swift)

**Comment by colasbd** [comment]
I'm using `SyntaxKit` to create an `NSTextView` that do syntax-highlight. Setting attributes leads to bugs whereas adding attributes is fine. We could also create two different functions (setting vs adding).

**Comment by alehed** [comment]
Additions to public API should always be well thought through.

The question is more how you are exactly planning to use the function to determine if adding it actually makes sense.

For instance we have the `parse` API which is very flexible and is expected to do the heavy lifiting in most cases. As a convenience function there is `attributedString` which facilitates the case really well where all you need is a static string to display in a view that is highlighted but not edited. So new API is really only desirable if it fits in one of two categories:
1. It does something that is very common with very little code (optimize for the common case)
2. It allows the user of the API something he could not do otherwise

Clearly we are looking at the first category here.
Now with this addition I'm not quite sure if it is either common enough or the only good way to achieve easy dynamic syntax highlighting. If I can be convinced that this is the case, then I am happy to merge it.

If you take a look at [TRexSyntaxKitDemo](https://github.com/dehlen/TRexSyntaxKitDemo) for instance, you can see the way it uses SyntaxKit is different. It first removes the attributes from the string. This might not make a difference since the callback gets called first with the global scope.

So here are the things that have to be clear:

* What is the exact use case? In a notification callback or the TextView delegate or somewhere else? Interactive highlighting or one time? Some code might make the intentions clear.
* Is this use case common? Assume the average user wants to highlight code im either a UITextView or an NSTextView.
* Is this the best way to satisfy the use case? Should we encourage the use case? How about incremental parsing etc?

**Comment by colasbd** [comment]
Hi @alehed,
Thanks for your questions. I think dynamic syntax highlighting is a very important use case of `SyntaxKit`! (unless there is a better library to do that... but so far, my choice went to `SyntaxKit`)

`UITextView` or `NSTextView` does not change anything. They both rely on `NSTextStorage` (which is a semi-concrete subclass of `NSMutableAttributedString`).

When some text is typed in a `NSTextView`, so many attributes are added to the string (tabs, colors, paragraphs, etc.). At first, I thought I won't care of these attributes. I just wanted the attributes given by `SyntaxKit` and the font. So `setAttributes:` was fine, I thought. But it didn't work at all (no text was displayed when I typed). That's why I ended choosing `addAttributes:`

I have not dug in the code of the `parser`, etc. Actually, I found the given examples not so easy to get. I remarked that you moved at out the API the `init` of the `Language` class and also `Theme` (I think). It tooks me some time to understand how to initialize these objects. In the branch your forked, the API was more flexible and I preferred it.

But my attention from the `README.md` was immediately drawn by this `attributedString(...)` convenience method.

I had a look at the library you mention (not `Swift 3` yet). I like their approach (thanks for the link). I will try it. A link to this project from the `README.md` would be a good idea.

Here is the code I use

```Swift
//
//  CBDLatexTextStorage.swift
//
//  Created by Colas Bardavid on 25/12/2016.
//  Copyright © 2016 Colas. All rights reserved.
//

import Foundation
import AppKit

@objc
public class CBDLatexTextStorage : NSTextStorage {
    
    private let storage = NSMutableAttributedString()
    private let latexParser: AttributedParser
    private let fontForLatex = NSFont.userFixedPitchFont(ofSize: 14)!
    
    // MARK: Init methods
    
    class func getLatexParser() -> AttributedParser {
        let bundle = Bundle(for: self)
        let bundleManager = BundleManager { (identifier, kind) -> (URL?) in
            let name = kind == .Language ? identifier._split(separator: ".")[1] : identifier
            let ext = kind == .Language ? ".tmLanguage" : ".tmTheme"
            return bundle.url(forResource: name.capitalized, withExtension: ext) ?? URL(fileURLWithPath: "")
        }
        
        let latex = bundleManager.language(withIdentifier: "source.Latex")!
        let tomorrow = bundleManager.theme(withIdentifier: "Tomorrow")!
        
        return AttributedParser(language: latex, theme: tomorrow)
    }
    
    override public init() {
        latexParser = CBDLatexTextStorage.getLatexParser()
        super.init()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        latexParser = CBDLatexTextStorage.getLatexParser()
        super.init(coder:aDecoder)
    }
    
    required public init?(pasteboardPropertyList propertyList: Any, ofType type: String) {
        latexParser = CBDLatexTextStorage.getLatexParser()
        super.init(pasteboardPropertyList: propertyList, ofType: type)
    }
    
    
    // MARK: Methods to override
    
    override public var string: String {
        return storage.string
    }
    
    override public func attributes(at location: Int, effectiveRange range: NSRangePointer?) -> [String : Any] {
        return storage.attributes(at: location, effectiveRange: range) as [String : AnyObject]
    }
    
    override public func replaceCharacters(in range: NSRange, with str: String) {
        self.beginEditing()
        storage.replaceCharacters(in: range, with: str)
        latexParser.addAttributes(to: storage)
        edited([.editedAttributes, .editedCharacters], range: range, changeInLength: str.characters.count - range.length)
        self.endEditing()
    }
    
    override public func setAttributes(_ attrs: [String : Any]?, range: NSRange) {
        self.beginEditing()
        storage.setAttributes(attrs, range: range)
        storage.addAttributes([NSFontAttributeName : fontForLatex], range: range)
        latexParser.addAttributes(to: storage)
        edited([.editedAttributes], range: range, changeInLength:0)
        self.endEditing()
    }
    
}
```

**Comment by colasbd** [comment]
I have tested both approaches. 
I have problem with TRexSyntaxKitDemo's approach. For now, I will keep mine.

**Comment by alehed** [comment]
Maybe it makes sense to provide this functionality. Is there a problem when you remove the foreground-color attributes before you re-apply them? Otherwise I would suggest providing a `reapplyAttributes`.

**Comment by alehed** [comment]
Note: I don't think you code needs to call `latexParser.addAttributes` in setAttributes.

**Comment by colasbd** [comment]
Thanks, I will test that.

**Comment by colasbd** [comment]
Yes, it is needed, when the text is loaded (from the model) for the first time. When typing it is not needed. But it is not needed in `replaceCharacters(...)` then.

Thanks

File: SyntaxKit/AttributedParser.swift [hunk 2]
```diff
@@ -43,6 +43,16 @@ open class AttributedParser: Parser {
         }
     }
 
+    open func addAttributes(to mutableString: NSMutableAttributedString) {
```
Comment (alehed) [inline SyntaxKit/AttributedParser.swift:46]: Maybe call it applyAttributes or applyHighlighting

File: SyntaxKit/AttributedParser.swift [hunk 3]
```diff
@@ -43,6 +43,16 @@ open class AttributedParser: Parser {
...
+            if let attributes = attributes {
+                mutableString.addAttributes(attributes, range: range)
+            }
+        }
+        mutableString.endEditing()
+    }
+    
     open func attributedString(for string: String, base: Attributes? = nil) -> NSAttributedString {
         let output = NSMutableAttributedString(string: string, attributes: base)
         output.beginEditing()
```
Comment (alehed) [inline SyntaxKit/AttributedParser.swift:58]: Use the applyAttributes function

File: SyntaxKit/AttributedParser.swift [hunk 4]
```diff
@@ -43,6 +43,16 @@ open class AttributedParser: Parser {
         }
     }
```
Comment (alehed) [inline SyntaxKit/AttributedParser.swift:45]: Maybe add a documentation comment that only the foreground attributes are changed

File: SyntaxKit/AttributedParser.swift [hunk 5]
```diff
@@ -43,6 +43,16 @@ open class AttributedParser: Parser {
...
 
+    open func addAttributes(to mutableString: NSMutableAttributedString) {
+        mutableString.beginEditing()
```
Comment (alehed) [inline SyntaxKit/AttributedParser.swift:47]: Does it make a difference if you remove all the the foreground attributes first with `removeAttribute()`?

**Review by alehed — CHANGES_REQUESTED** [review]
Just a few comments below. I thought about it and even though it is not recommended to re-parse everything for interactive use, a function that re-colors the string is probably desirable.

**Comment by colasbd** [comment]
I won't have access to my computer until next week but then I'll review your comments.

Actually, I'm thinking of a different strategy: instead of subclassing NSTextStorage use a dedicated method of NSLayoutManger to addAttributes that don't affect the layout.

And because of SyntaxKit being too slow, I'm considering studying the code of TexShop. It does syntax highlighting for LaTeX and might be quicker.

**Comment by alehed** [comment]
If you just have to highlight one language like latex and you are not afraid of parsers (trust me, there is nothing scary about them) just write a recursive descent parser yourself or have one generated with a parser generator like Antlr. That is the fastest and most exact method to recognize and color syntax.

You might even find a good unambiguous pre-made grammar for tex on the internet.

**Comment by colasbd** [comment]
Finally, I used another project to parse and colorize .tex code.
I will not use SyntaxKit.

Thanks for your help :+1: :-)

**Comment by alehed** [comment]
Ok, I'm sure there are lots of libraries that are faster than SyntaxKit. 

But that's no reason to close all the issues. If you don't like to be bothered by the issues any more, you can simply turn off notifications.

**Comment by colasbd** [comment]
Sorry :(
Do you want me to reopen the PR?

I meant to be polite by closing the issue since I opened them and
1) the issue with LaTeX likely comes from the need to include several `.tmLanguage`
2) there is no issue with speed (a better documentation would be great)

**Comment by alehed** [comment]
Thanks! 👍

The PR is fine, you don't have to reopen that.

I still have to check on the highlighting of `\documentclass` and it really is slow. Now I don't know what I can do about it (except tell people to use asynchronous incremental highlighting) but it is good to know that the issue is there for other people investigating whether they should use the project.

**Pull request closed without merging**